Convert entire source tree from json_spirit to UniValue

jonasschnelli opened the pull request:
Takes up and supersedes #4738.

This PR keeps the JSON Spirit "backward-compatibility" and **passes all tests**. Therefore there is a new `UniValue` number type `VREAL` to distinct between int and double with the current JSON spirit precision of 8 digits after point.
Because after rfc4627 (and UniValue honors that) some basic types/strings like `"null"` or `"0"` and `"string"` are not valid JSON strings, i had to slightly change some unit tests (will comment below directly in the code).

Commit:
UniValue: prefer .size() to .count(), to harmonize w/ existing tree

Commit:
UniValue: export NullUniValue global constant

Commit:
Convert tree to using univalue. Eliminate all json_spirit uses.

jonasschnelli committed:
extend conversion to UniValue

jonasschnelli committed:
expicit set UniValue type to avoid empty values

jonasschnelli committed:
special threatment for null,true,false because they are non valid json

jonasschnelli committed:
univalue: add support for real, fix percision and make it json_spirit compatible

- avoid breaking the API because of different number/percision handling

jonasschnelli committed:
univalue: correct bool support

jonasschnelli committed:
fix rpc unit test, plain numbers are not JSON compatible object

UniValues read() does only read valid json.

jonasschnelli commented on src/bitcoin-cli.cpp:
```diff
@@ -142,8 +142,8 @@ Object CallRPC(const string& strMethod, const Array& params)
     // Parse reply
-    Value valReply;
-    if (!read_string(strReply, valReply))
+    Value valReply(UniValue::VSTR);
```
UniValue required explicit type definition during constructing.
`Value` should be replaced with `UniValue` soon.

jonasschnelli commented:
I did use merge instead of rebase to clearly see my adaptations from #4738.

jonasschnelli commented on src/rpcclient.cpp:
```diff
@@ -134,9 +136,20 @@ Array RPCConvertValues(const std::string &strMethod, const std::vector<std::stri
+                if (!jVal.read(strVal) || (jVal.isNull() && strVal.size() > 0))
+                    if(!jVal.setNumStr(strVal) || jVal.isNull())
+                        throw runtime_error(string("Error parsing JSON:")+strVal);
+            }
```
I had to extend this to gain compatibility between JSON Spirit and UniValue. UniValue is more strict in case of JSON encoding.

jonasschnelli commented on src/test/rpc_tests.cpp:
```diff
@@ -111,20 +111,20 @@ BOOST_AUTO_TEST_CASE(rpc_rawsign)
-    Value value;
-    BOOST_CHECK(read_string(str, value));
+    UniValue value;
+    BOOST_CHECK(value.setNumStr(str));
```
Had to change this to `setNumStr` because `read()` only parses valid JSON (things like "0.00000001" is not valid JSON).

jonasschnelli commented on src/test/univalue_tests.cpp:
```diff
@@ -49,7 +49,7 @@ BOOST_AUTO_TEST_CASE(univalue_constructor)
     double vd = -7.21;
     UniValue v7(vd);
-    BOOST_CHECK(v7.isNum());
+    BOOST_CHECK(v7.isReal());
```
Because now there is a distinction between int and real (to make it JSON Spirit compatible), things like -7.21 are VREAL and not VNUM. I decided to make VREAL a standalone type instead of a VNUM subtype.

jonasschnelli commented on src/test/univalue_tests.cpp:
```diff
@@ -240,11 +240,11 @@ BOOST_AUTO_TEST_CASE(univalue_object)
 
 static const char *json1 =
-"[1.1,{\"key1\":\"str\",\"key2\":800,\"key3\":{\"name\":\"martian\"}}]";
+"[1.10000000,{\"key1\":\"str\",\"key2\":800,\"key3\":{\"name\":\"martian\"}}]";
```
To keep the JSON/RPC API backward compatible, all reals represented in JSON have a fixed precision to 8 digits after point.

jonasschnelli commented on src/univalue/univalue.h:
```diff
@@ -26,6 +30,9 @@ class UniValue {
     }
+    UniValue(bool val_) {
+        setBool(val_);
+    }
```
This PR also adds a clean boolean handling to UniValue (necessary for JSONRPC compatibility).

jgarzik commented:
ut ACK - reviewed the whole thing

Will do some testing later this week.

Thanks for moving this forward!

laanwj commented:
Thanks for doing this work.

One remark: I see `AmountFromValue` and `ValueFromAmount` still make the value pass through a `double.` My prime gripe with JSON spirit is that monetary values still had to be converted from and to floating point which can cause deviations (see See #3759 and https://bitcoin.stackexchange.com/questions/22716/bitcoind-sendfrom-round-amount-error ). 

Ideally a new JSON implementation will avoid this and interpret Numbers from/to fixed-point or decimal directly. Similar to Python's 

    postdata = json.dumps(list(rpc_call_list), default=EncodeDecimal)
    response = json.loads(responsedata, parse_float=decimal.Decimal)

jgarzik commented:
Agree w/ your gripe @laanwj.  However, I think of that as a separate logical step:
1. Convert tree with as much equivalence as possible, to prove everything works as it always did.
2. Improve the tree beyond json-spirit equivalence.

Separated thusly it is easier to test the conversion before moving on to more ambitious cleanups.

jonasschnelli committed:
remove JSON Spirit UniValue wrapper

jonasschnelli committed:
Remove JSON Spirit wrapper, remove JSON Spirit leftovers

- implement find_value() function for UniValue
- replace all Array/Value/Object types with UniValues, remove JSON Spirit to UniValue wrapper
- remove JSON Spirit sources

jonasschnelli commented:
This is now complete.

Just added another commit (https://github.com/bitcoin/bitcoin/commit/2a3126c336fc0cc2b19f9b89bd48ed3e489a0817) which does completely remove JSON Spirit (also the UniValue wrapper) from the sources and the compile process.

Passes all tests.

If the diff size makes this PR to risky, there would also be the chance to pull this in without https://github.com/bitcoin/bitcoin/commit/2a3126c336fc0cc2b19f9b89bd48ed3e489a0817 or / and separate https://github.com/bitcoin/bitcoin/commit/2a3126c336fc0cc2b19f9b89bd48ed3e489a0817 into another PR.

I think it would be good to remove JSON Spirit in this PR while not breaking the RPC API. After this has been done, we could focus on optimizing the monetary values.

jonasschnelli commented:
Gitian was also happy with this: http://builds.jonasschnelli.ch/pulls/6121/

jgarzik commented:
Yes, I agree JSON-spirit should be removed in this PR.

Reviewing & testing right now...

jonasschnelli commented:
Rebased (puhh).
Thanks in advance for reviewing this (maybe @jgarzik and @theuni) and finally bring this forward.

laanwj commented:
(apparantly) spurious travis failure with the comparison tool, retriggering

laanwj commented:
Looks good to me. Tested and reviewed ACK.
Would like to merge this asap now that 0.11 is branched.

laanwj commented:
This breaks `linearize-hashes.py`. Likely, this is caused by JSON-RPC batching no longer working:
```
./linearize-hashes.py  linearize.cfg 
Traceback (most recent call last):
  File "./linearize-hashes.py", line 112, in <module>
    get_block_hashes(settings)
  File "./linearize-hashes.py", line 68, in get_block_hashes
    for x,resp_obj in enumerate(reply):
TypeError: 'NoneType' object is not iterable
```
(expected output: a list of block hashes)

jonasschnelli commented:
Thanks for testing. Nice catch!
Will have a close look at the batching feature within the next hours.

jonasschnelli committed:
fix rpc batching univalue issue

jonasschnelli commented:
RPC batching issue is fixed.

laanwj commented:
Can confirm that it is fixed.

jonasschnelli committed:
fix missing univalue types during constructing

jonasschnelli commented:
added another commit that fixes a univalue type issue within `getrawmempool` and `getmempoolinfo`

jtimon commented:
Concept ACK. 
Is this getting into 0.11?
This kind of change tends to be painful to maintain...

jonasschnelli commented:
@jtimon It's not for 0.11. It's meant for current master which aims for 0.12.

laanwj commented:
Needs rebase after #6226, will merge after that

laanwj committed:
Simplify RPCclient, adapt json_parse_error test

# Conflicts:
#	src/test/rpc_tests.cpp

jonasschnelli committed:
fix univalue json parse tests

laanwj committed:
univalue: add strict type checking

laanwj committed:
util: Add ParseInt64 and ParseDouble functions

Strict parsing functions for other numeric types.

- ParseInt64 analogous to ParseInt32, but for 64-bit values.
- ParseDouble for doubles.
- Make all three Parse* functions more strict (e.g. reject whitespace on
  the inside)

Also add tests.

jonasschnelli committed:
univalue: add type check unit tests

sdaftuar commented:
I'm getting an error now when running ```qa/rpc-tests/getblocktemplate_proposals.py```:
```JSONRPC error: Missing data String key for proposal```

Looks like it's here:
```
File "./getblocktemplate_proposals.py", line 86, in assert_template
    rsp = node.getblocktemplate({'data':template_to_hex(tmpl, txlist),'mode':'proposal'})
```

jonasschnelli commented:
Meh. This bug crept in during multiple transitions from json spirit objects to UniValue.
Fixed with #6234.

Thanks @sdaftuar for reporting!

jgarzik commented:
Odd that I missed that in my testing.  Anyway, re-ACK!  Tested.

theuni commented:
@jonasschnelli Sorry for not getting to a final review in time. I'll still look over it again in master for good measure. Great work!
